turtlebot2_drivers/src/kobuki_node.cpp

```diff
@@ -22,76 +22,128 @@
+
+void cmdVelCallback(const geometry_msgs::msg::Twist::SharedPtr msg) {
   std::lock_guard<std::mutex> kobuki_guard(g_kobuki_mutex);
-  std::cout << "Received: (" << msg->linear.x << "," << msg->angular.z << ")" << std::endl;
+  printf("Received: (%6.3f, %6.3f)\n", msg->linear.x, msg->angular.z);
   double vx = std::min(std::max(msg->linear.x, -g_max_vx), g_max_vx);
   double vyaw = std::min(std::max(msg->angular.z, -g_max_vyaw), g_max_vyaw);
   g_kobuki->setBaseControl(vx, vyaw);
   g_last_cmd_vel_time = std::chrono::system_clock::now();
 }
 
-void on_parameter_event(const rcl_interfaces::msg::ParameterEvent::SharedPtr event)
-{
-  // TODO: handle dynamic params
-  std::cout << "Parameter event:" << std::endl << " new parameters:" << std::endl;
-  for (auto & new_parameter : event->new_parameters) {
-    std::cout << "  " << new_parameter.name << std::endl;
-  }
-  std::cout << " changed parameters:" << std::endl;
-  for (auto & changed_parameter : event->changed_parameters) {
-    std::cout << "  " << changed_parameter.name << std::endl;
+rcl_interfaces::msg::SetParametersResult on_param_change(
+    const std::vector<rclcpp::parameter::ParameterVariant> parameters) {
+  rcl_interfaces::msg::SetParametersResult result;
+  result.successful = true;
+
+  double max_vx = g_max_vx;
+  double max_vyaw = g_max_vyaw;
+  std::string port = k_parameters.device_port;
+
+  for (auto param : parameters) {
+    if ("device_port" == param.get_name() &&
+        param.get_type() ==
+            rcl_interfaces::msg::ParameterType::PARAMETER_STRING) {
+      if (nullptr == g_kobuki) {
+        port = param.get_value<std::string>();
+      } else {
+        result.successful = false;
+        result.reason = "Dynamically changing device_port is not supported, "
+                        "using old value";
+      }
+    } else if ("device_port" == param.get_name() &&
+               param.get_type() !=
+                   rcl_interfaces::msg::ParameterType::PARAMETER_STRING) {
+      result.successful = false;
+      result.reason = "device_port has to be a string";
+    }
+    if ("max_vx" == param.get_name() &&
+        param.get_type() ==
+            rcl_interfaces::msg::ParameterType::PARAMETER_DOUBLE) {
+      max_vx = param.get_value<double>();
+    } else if ("max_vx" == param.get_name() &&
+               param.get_type() !=
+                   rcl_interfaces::msg::ParameterType::PARAMETER_DOUBLE) {
+      result.successful = false;
+      result.reason = "max_vx has to be a double";
+    }
+    if ("max_vyaw" == param.get_name() &&
+        param.get_type() ==
+            rcl_interfaces::msg::ParameterType::PARAMETER_DOUBLE) {
+      max_vyaw = param.get_value<double>();
+    } else if ("max_vyaw" == param.get_name() &&
+               param.get_type() !=
+                   rcl_interfaces::msg::ParameterType::PARAMETER_DOUBLE) {
+      result.successful = false;
+      result.reason = "max_vyaw has to be a double";
+    }
```

Comment:
These repeated conditions are hard to read. I would suggest to extract the common conditions into a first level and then use a condition and an else block for the second level, e.g.:

```
if ("device_port" == param.get_name()) {
  if (param.get_type() == rcl_interfaces::msg::ParameterType::PARAMETER_STRING) {
    ...
  } else {
    ...
  }
}
```